File: src/main/java/ru/AlexeyFedechkin/znatoki/StudyRussianBot/TelegramBot.java
package ru.AlexeyFedechkin.znatoki.StudyRussianBot;

import org.apache.log4j.Logger;
import org.telegram.telegrambots.bots.DefaultBotOptions;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.ApiContext;
import org.telegram.telegrambots.meta.TelegramBotsApi;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException;
import ru.AlexeyFedechkin.znatoki.StudyRussianBot.Object.User;

public class TelegramBot extends TelegramLongPollingBot {

    private final Logger logger = Logger.getLogger(TelegramBot.class);
    private Parser parser;

    private TelegramBot(DefaultBotOptions botOptions) {
        super(botOptions);
    }

    public TelegramBot(){
    }

    /**
     * init telegram bot and configure proxy
     */
    public void init(){
//        Authenticator.setDefault(new Authenticator() {
//            @Override
//            public PasswordAuthentication requestPasswordAuthenticationInstance(String host, InetAddress addr,
//                                                                                int port, String protocol,
//                                                                                String prompt, String scheme, URL url,
//                                                                                RequestorType reqType) {
//                return new PasswordAuthentication(Config.getInstance().getProxyUser(),
//                        Config.getInstance().getProxyPassword().toCharArray());
//            }
//        });
        TelegramBotsApi botsApi = new TelegramBotsApi();
        DefaultBotOptions botOptions = ApiContext.getInstance(DefaultBotOptions.class);
        botOptions.setProxyHost(Config.getInstance().getProxyHost());
        botOptions.setProxyPort(Config.getInstance().getProxyPort());
        botOptions.setProxyType(DefaultBotOptions.ProxyType.SOCKS5);
        logger.info("proxy configure");
        try {
            botsApi.registerBot(new TelegramBot(botOptions));
            logger.info("bot register");
        } catch (TelegramApiRequestException e) {
            logger.fatal("bot start fail", e);
            System.exit(20);
        }
    }

    /**
     * @param update
     */
    public void onUpdateReceived(Update update) {
        if (parser == null){
            parser = new Parser(this);
        }
        if (update.hasCallbackQuery()){
            if (!parser.getUsers().containsKey(update.getCallbackQuery().getMessage().getChatId())){
                parser.getUsers().put(update.getCallbackQuery().getMessage().getChatId(), new User(update.getCallbackQuery().getMessage().getChatId()));
            }
        } else {
            if (!parser.getUsers().containsKey(update.getMessage().getChatId())){
                parser.getUsers().put(update.getMessage().getChatId(), new User(update.getMessage().getChatId()));
            }
        }
        if (update.hasCallbackQuery()){
            logger.info("receive callback " + update.getCallbackQuery().getData() + " " +
                    update.getCallbackQuery().getFrom().getFirstName() + " " +
                    update.getCallbackQuery().getFrom().getLastName() + " " +
                    update.getCallbackQuery().getFrom().getUserName());
            parser.parsCallback(update);
        } else {
            logger.info("receive message " + update.getMessage().getText() +
                    " from " + update.getMessage().getFrom().getFirstName() + " " +
                    update.getMessage().getFrom().getLastName() + " " +
                    update.getMessage().getFrom().getUserName());
            parser.parseText(update);
        }
    }

    /**
     * @param message string with message to user
     * @param chatId id of chat where to send message
     * @return result of sending
     */
    public boolean sendMessage(String message, long chatId){
        logger.info("send message " + message);
        SendMessage msg = new SendMessage().
                setChatId(chatId).
                setText(message);
        try{
            execute(msg);
            return true;
        } catch (TelegramApiException e) {
            logger.warn("fail to send message", e);
            return false;
        }
    }

    public void sendMessage(SendMessage sendMessage){
        try{
            execute(sendMessage);
        } catch (TelegramApiException e) {
            logger.warn("fail to send message", e);
        }
    }

    /**
     * @return
     */
    public String getBotUsername() {
        logger.info("getting bot user name");
        return Config.getInstance().getBotUserName();
    }

    /**
     * @return
     */
    public String getBotToken() {
        logger.info("getting bot token");
        return Config.getInstance().getBotToken();
    }
}

File: src/main/java/me/centralhardware/znatoki/studyRussianBot/telegram/TelegramBot.kt
package me.centralhardware.znatoki.studyRussianBot.telegram

import kotlinx.coroutines.runBlocking
import me.centralhardware.znatoki.studyRussianBot.Config
import me.centralhardware.znatoki.studyRussianBot.objects.User
import org.slf4j.LoggerFactory
import org.telegram.telegrambots.longpolling.TelegramBotsLongPollingApplication
import org.telegram.telegrambots.longpolling.util.LongPollingSingleThreadUpdateConsumer
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException
import kotlin.system.exitProcess


/**
 *telegram bot class
 */
class TelegramBot : LongPollingSingleThreadUpdateConsumer {
    private var telegramParser: TelegramParser? = null
    private var inlineKeyboard: InlineKeyboard
    private var sender: Sender = Sender()

    init {
        inlineKeyboard = InlineKeyboard(sender)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(TelegramBot::class.java)

        /**
         * init telegram bot and configure proxy
         */
        fun init() {
            try {
                val application = TelegramBotsLongPollingApplication()
                application.registerBot(Config.token, TelegramBot())
                logger.info("bot register")
            } catch (e: TelegramApiRequestException) {
                logger.warn("bot start fail", e)
                exitProcess(20)
            }
        }
    }

    /**
     * method by which the library telegram sends
     * the received messages for processing by the server part
     * logging input message.
     * @param update received message
     */
    override fun consume(update: Update){
        runBlocking {
            try {
                process(update)
            } catch (t: Throwable) {
                logger.warn("", t)
            }
        }
    }

    private suspend fun process(update: Update) {
        if (telegramParser == null) telegramParser = TelegramParser(sender)

        when {
            update.hasCallbackQuery() -> {
                logger.info(
                    "receive callback \"" + update.callbackQuery.data + "\" " +
                            update.callbackQuery.from.firstName + "\" " +
                            update.callbackQuery.from.lastName + "\" " +
                            update.callbackQuery.from.userName + "\""
                )
                if (!telegramParser!!.users.containsKey(update.callbackQuery.message.chatId)) {
                    telegramParser!!.users[update.callbackQuery.message.chatId] =
                        User(update.callbackQuery.message.chatId)
                }
                telegramParser!!.parsCallback(update)
            }

            update.message.hasText() -> {
                logger.info(
                    "receive message \"" + update.message.text +
                            "\" from \"" + update.message.from.firstName + "\" " +
                            update.message.from.lastName + "\" " +
                            update.message.from.userName + "\" " +
                            update.message.from.id + "\""
                )
                when {
                    !telegramParser!!.users.containsKey(update.message.chatId) ->
                        telegramParser!!.users[update.message.chatId] = User(update.message.chatId)
                }
                telegramParser!!.parseText(update)
            }
        }
    }
}